Add support for user info REST API

austinzheng opened the pull request:
Changes:
- Added a `RLMSyncUserInfo` class to encapsulate info retrieved from a server about a user
- Added an API for admin users to retrieve info about a user
- Refactored the internal JSON network client to enable the new endpoint
- Consolidated all the internal JSON model classes in the same set of source files
- Defined a few new internal key strings

Fixes #5060

TODO:
- [x] Resolve the TODO comments
- [x] Write tests (needs a ROS with this feature)
- [x] Changelog entry

This is _not_ a breaking change.

austinzheng commented:
This is ready for review. I know the test for the new feature passes but I'm still having trouble running the entire object server test suite on my machine, so I'll be closely watching the result of the equivalent CI job.

bdash commented:
@jpsim, can you please review this when you have a moment?

jpsim commented on Realm/ObjectServerTests/RLMObjectServerTests.mm:
```diff
@@ -388,7 +388,73 @@ - (void)testOtherUserChangePassword {
+    [testUser retrieveUserInfoForProvider:RLMIdentityProviderUsernamePassword
+                         providerIdentity:@"invalid_user@realm.example.org"
+                               completion:^(RLMSyncUserInfo *info, NSError *err) {
+                                   XCTAssertNotNil(err);
```
Should we validate that the error matches what we expect here? Code, message, etc?

jpsim commented on Realm/ObjectServerTests/RLMObjectServerTests.mm:
```diff
@@ -388,7 +388,73 @@ - (void)testOtherUserChangePassword {
+    [user1 retrieveUserInfoForProvider:RLMIdentityProviderUsernamePassword
+                         providerIdentity:username2
+                               completion:^(RLMSyncUserInfo *info, NSError *err) {
+                                   XCTAssertNotNil(err);
```
Should we validate that the error matches what we expect here? Code, message, etc?

jpsim commented on Realm/ObjectServerTests/RLMObjectServerTests.mm:
```diff
@@ -388,7 +388,73 @@ - (void)testOtherUserChangePassword {
+
+    // Create a non-admin user.
+    RLMSyncCredentials *c1 = [RLMSyncCredentials credentialsWithUsername:username1 password:pw register:YES];
+    RLMSyncUser *user1 = [self logInUserForCredentials:c1 server:server];
```
Can we use more descriptive variable names? `user1` -> `nonAdminUser`, `user2` -> `adminUser1`, `testUser` -> `adminUser2`.

bdash commented on Realm/RLMSyncUser.h:
```diff
@@ -173,6 +179,22 @@ NS_SWIFT_UNAVAILABLE("Use the full version of this API.");
+ */
+- (void)retrieveUserInfoForProvider:(RLMIdentityProvider)provider
+                   providerIdentity:(NSString *)providerID
+                         completion:(RLMRetrieveUserBlock)completion;
```
I'm not sure I like this naming. `providerIdentity` suggests to me that parameter is the identity of the provider, when in fact it's the provider's concept of the identity of the user. Perhaps `-retrieveInfoForUser:identityProvider:completion:`?

austinzheng replied:
Yeah, that looks a lot more natural.

bdash commented on Realm/RLMSyncUser.h:
```diff
@@ -245,6 +267,41 @@ NS_SWIFT_UNAVAILABLE("Use the full version of this API.");
+@property (nonatomic, readonly) NSString *providerID;
+
+/**
+ The internal Realm Object Server identity of the user.
```
What do we mean by "internal" here? It sorta has the implication of "private", which obviously isn't right.

bdash commented on Realm/RLMSyncUser.h:
```diff
@@ -245,6 +267,41 @@ NS_SWIFT_UNAVAILABLE("Use the full version of this API.");
+/**
+ The provider-based username of the user.
+ */
+@property (nonatomic, readonly) NSString *providerID;
```
Is this a username or identity? The doc comment suggests a username, but the property name indicates otherwise. It'd also be good to be consistent about whether we use `identity` or `ID` in these property names. We're using both within this same class.

austinzheng replied:
I'm just going to call it `username`, `identity` and related terms are too confusing.

bdash commented:
It looks like this needs further revision to conform to the latest API consensus.

austinzheng commented:
I'll fix it up first thing tomorrow and comment again when it's ready for (re-)review.

austinzheng commented:
I made the changes, squashed everything, and rebased onto latest master to get rid of GitHub's merge conflict complaints. It should be ready for final review.

bdash commented:
The API now looks consistent with the consensus we'd arrived at. 👍

@jpsim, can you please do another review pass on this?

austinzheng commented:
Fixed object server test failure. Rebased onto latest master and manually resolved merge conflicts.

jpsim commented on RealmSwift/Sync.swift:
```diff
@@ -23,11 +23,19 @@ import Foundation
  An object representing a Realm Object Server user.
 
- - see: `SyncUser`
+ - see: `RLMSyncUser`
```
why this change?
